### ISSUE TITLE
BB-29 - [BUG] Issues with Debian preseed

### ISSUE THREAD

**Issue by erkrali** [bluebanquise/bluebanquise#894]
Hi,

The Debian preseed file has a couple issues:

The main one is that the final script (`d-i preseed/late_command`) is missing some newline escaping (the `\` at the end of the line), so the script gets interrupted. Seen at least [here](https://github.com/bluebanquise/bluebanquise/blob/master/collections/infrastructure/roles/pxe_stack/templates/Debian/preseed.cfg.j2#L100), [here](https://github.com/bluebanquise/bluebanquise/blob/master/collections/infrastructure/roles/pxe_stack/templates/Debian/preseed.cfg.j2#L102) and [here](https://github.com/bluebanquise/bluebanquise/blob/master/collections/infrastructure/roles/pxe_stack/templates/Debian/preseed.cfg.j2#L110). This made that the default iPXE boot option didn't switch to "local", and you would end up with an install-loop.

The other one is that it would be really nice to have some points to hook personalised options. For instance, you [hardcode the http proxy to empty](https://github.com/bluebanquise/bluebanquise/blob/master/collections/infrastructure/roles/pxe_stack/templates/Debian/preseed.cfg.j2#L29), but my nodes have no direct internet access and thus need to use a proxy. Or perhaps somebody wants to use a different, possibly local, mirror (that's also hardcoded).

Now this is rather a feature request, and I'm not sure what would be the best approach here: either let the users override the template and provide their own template via some ansible variable, or just add some "hooks" where the user can add snippets. In fact, the partitioning snippet can be (ab)used for this purpose, but it feels less clean, and some options might be later overridden.

An interesting option would be [chaining preseed files](https://www.debian.org/releases/stable/i386/apbs05.en.html#preseed-chainload), but as I said, I don't know how that would integrate in the overall design.

As a final note, you can also add preseed options through the kernel command line (which I did), but they get overridden later if present in the preseed file (for instance the http proxy).

**Comment by oxedions** [comment]
Dear @erkrali 

Many thanks for this second issue :blush: 
This is very interesting !

I think I introduced the new line bug with the recent fix for uid/gid, I will check asap.

Ok for proxy, you are the second person to ask this month, I will add it for all distributions too because it seems a requirement for many sites.

Regarding the feature request, this is something very feasible yes. Basically, I can do something like this:

```
{% if os_autoinstall_raw_file is defined %}
{{ os_autoinstall_raw_file }}
{% else %}
stuff
foo
bar
  {% if os_autoinstall_raw_content is defined %}
{{ os_autoinstall_raw_content }}
  {% endif %}
{% endif %}
```

So you can either overwrite the whole preseed file with your own content, and still benefit from the BlueBanquise pxe mechanism, or just add some content to the existing template. Would that be what you are looking for ?

I didn't know preseed could chain. This is interesting. I need to check if other distributions can do that too (kickstart, ubuntu curtin, and autoyast).

**Comment by erkrali** [comment]
I think the proxy feature really makes sense, because in my experience, in many environments only the head nodes (and eventually the login nodes) get external access, whereas the "worker" nodes just get an internal network and a proxy. BTW, you might eventually consider adding a `squid` role eventually (that's a feature request within a feature request :grimacing:)

This is usually done to prevent a bunch of nodes (can be 100s or even 1000s in big clusters) from hitting the same poor unexpecting server at around the same time and causing an (unintentional) DoS :smile:

About your solution for customization, it looks really cool. If done like that, you wouldn't really need the chaining, because the user would be free to just add it as needed in the `{{ os_autoinstall_raw_content }}` part. This way one wouldn't need to put the full content in an ansible variable, and it can even be managed externally.

Of course, you can also use it internally to keep things cleaner (different sections of the preseed / kickstart / etc.) At least kickstart seems to have something similar ([`%include`](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/system_design_guide/kickstart-commands-and-options-reference_system-design-guide#include_kickstart-commands-for-installation-program-configuration-and-flow-control) and [`%ksappend`](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/system_design_guide/kickstart-commands-and-options-reference_system-design-guide#ksappend_kickstart-commands-for-installation-program-configuration-and-flow-control)).

**Comment by oxedions** [comment]
Dear @erkrali 

Thanks a lot for this feedback :blush: 
Ok for proxy then. Its simple to setup.

Hum, the squid role makes a lot of sens indeed. I will investigate this. It would be nice as a proxy for clusters that uses external repositories for packages directly. Even on my infra I would be happy to have it, instead of having to re-download same packages all the time.

I will work on this and finish (already started) the fix for DNS after the 5th of march. I will be off during 1 week, my family needs it.
Apologies for the delay :slightly_smiling_face:

**Comment by erkrali** [comment]
Hi!

After some more experimenting, I found another bug  [here](https://github.com/bluebanquise/bluebanquise/blob/master/collections/infrastructure/roles/pxe_stack/templates/Debian/preseed.cfg.j2#L112).

In short, the `in-target` command stops at the first `;`, so the `do` ends up hanging. Also, if you do it `in-target`, I'm not sure it would propagate to future `in-target` invocations (if they start in a new shell), and furthermore `/proc` might not be available.

So, in short, I just ended up removing `in-target` and leaving it like this:

```
  for I in `cat /proc/cmdline | tr " " "\n" | grep -E "^(node_hostname|ipxe_next_server)="` ; do eval $I; done; \
```

**Comment by erkrali** [comment]
In fact, the line isn't even necessary, the kernel will pass any unrecognized parameter as an environment variable:

> Anything of the form 'foo=bar' that is not accepted as a setup function as described above is then interpreted as an environment variable to be set.  A (useless?) example would be to use 'TERM=vt100' as a boot argument.

https://man7.org/linux/man-pages/man7/bootparam.7.html

**Comment by oxedions** [comment]
Dear @erkrali 

Many thanks and apologies for the delay of my answer :blush: 

This is very interesting, I will check this because this would simplify a lot this line. I will fix that asap :slightly_smiling_face:

**Comment by oxedions** [comment]
Starting working on this issue.
- [ ] Add proxy support for auto installation files (all OS)
- [ ] Add raw content support for auto installation files (all OS)
- [ ] Add full file raw content support for auto installation files (all OS)
- [ ] Fix preseed late commands (see all errors described above) (Debian only)
- [ ] Consider adding a squid role

Note: beware of Ubuntu curtin format, indenting could be an issue with raw content. Check if curtin supports merged by same dicts keys or not to solves that. If not, no solution.

**Comment by oxedions** [comment]
> In fact, the line isn't even necessary, the kernel will pass any unrecognized parameter as an environment variable:
> 
> > Anything of the form 'foo=bar' that is not accepted as a setup function as described above is then interpreted as an environment variable to be set.  A (useless?) example would be to use 'TERM=vt100' as a boot argument.
> 
> https://man7.org/linux/man-pages/man7/bootparam.7.html

I did a small test on RHEL 9:

I removed the 
`for I in `cat /proc/cmdline | tr " " "\n" | grep -E "^(node_hostname|ipxe_next_server)="` ; do eval $I; done`

And only kept the:
`curl -s -k http://$ipxe_next_server/cgi-bin/bootswitch.cgi --data "node=$node_hostname&boot=disk"`

But it didnt work, at the end in logs, I can see it only did:

```
[root@mgt3 ~]# cat bluebanquise_main_post-install.log 
+ sed -i 's/^.*PermitRootLogin.*$/PermitRootLogin\ no/' /etc/ssh/sshd_config
+ mkdir /var/lib/bluebanquise/.ssh
+ cat
+ restorecon -R -v /var/lib/bluebanquise/.ssh
+ echo 'bluebanquise ALL=(ALL:ALL) NOPASSWD:ALL'
+ echo 'All went well, we can inform next-server we succeeded'
All went well, we can inform next-server we succeeded
+ curl -s -k http:///cgi-bin/bootswitch.cgi --data 'node=&boot=disk'
+ set +x
[root@mgt3 ~]#
```

I will check. I know RedHat kernel are heavily patched, so maybe there is something I miss here. Investigating.

**Comment by erkrali** [comment]
I think it's not about the kernel, but rather the init process.

The environment variables go to the init process, but it doesn't meen that they are passed down to children processes.

 [This](https://utcc.utoronto.ca/~cks/space/blog/linux/KernelCmdlineProcessing) is a bit old, bit it seems to explain it:

> For anything that is not a kernel option and does not have a '.' in its name (these are assumed to be unhandled module parameters), one of two things happen. If it is of the form 'a=b' or (I believe) 'a=', it's placed in the environment that will be passed to the initial user-level process (generally either /init from the initial ramfs or /sbin/init). If it doesn't look like an environment variable setting it becomes one of the command line arguments for the initial user-level process. Everything from the kernel command line also appears in /proc/cmdline.

> Fedora 17's /init appears to completely ignore command line arguments and does not pass them to systemd's /sbin/init.

It seems RHEL init system just not passes them on.

**Comment by oxedions** [comment]
Dear @erkrali

You can find current fixes here : https://github.com/bluebanquise/bluebanquise/pull/911/

Could you have a quick look ?

I am working on the squid role, this should be a simple one (as a starter, more features can be added later).

**Comment by oxedions** [comment]
> I think it's not about the kernel, but rather the init process.
> 
> The environment variables go to the init process, but it doesn't meen that they are passed down to children processes.
> 
> [This](https://utcc.utoronto.ca/~cks/space/blog/linux/KernelCmdlineProcessing) is a bit old, bit it seems to explain it:
> 
> > For anything that is not a kernel option and does not have a '.' in its name (these are assumed to be unhandled module parameters), one of two things happen. If it is of the form 'a=b' or (I believe) 'a=', it's placed in the environment that will be passed to the initial user-level process (generally either /init from the initial ramfs or /sbin/init). If it doesn't look like an environment variable setting it becomes one of the command line arguments for the initial user-level process. Everything from the kernel command line also appears in /proc/cmdline.
> 
> > Fedora 17's /init appears to completely ignore command line arguments and does not pass them to systemd's /sbin/init.
> 
> It seems RHEL init system just not passes them on.

Ok thanks a lot ! This is very interesting. I did not add the kernel parameters for this OS like for the other ones.

**Comment by erkrali** [comment]
> Dear @erkrali
> 
> You can find current fixes here : #911
> 
> Could you have a quick look ? There are probably still some stupid bugs I made, I am working on it. But its just for you to check the concept would answer your needs 😊
> 
> I am working on the squid role, this should be a simple one (as a starter, more features can be added later).

I will have a look, but the idea looks really well thought-out! Thanks for the great work!

**Comment by oxedions** [comment]
Dear @erkrali 

I merged https://github.com/bluebanquise/bluebanquise/pull/911
This contains all what we discussed so far here :blush: 

It should (I hope) answer all your needs on this topic. Do not hesitate if something is not working as you expected !

**Comment by erkrali** [comment]
Hi again!

Sorry, I got busy with other things and then came Easter, so I didn't have time to test it.

I have just tested it, and I have found a very minor bug:

In [the documentation](https://github.com/bluebanquise/bluebanquise/blob/74aa35920a6dbc949402e4cdef2e1d73570034a4/collections/infrastructure/roles/pxe_stack/README.md#proxy-during-pxe), you mention two different variables, `os_pxe_repository_proxy` and `os_pxe_proxy`.

At least for Debian, the latter should be used as a kernel parameter. However, the [relevant code](/https://github.com/bluebanquise/bluebanquise/blob/74aa35920a6dbc949402e4cdef2e1d73570034a4/collections/infrastructure/roles/pxe_stack/templates/Debian/equipment_profile_preseed.ipxe.j2#L16) shows that you're also using `os_pxe_repository_proxy`.

In fact, I grepped for `os_pxe_proxy` in the whole repository, and I could only find it in the documentation, so you might want to double check it for other systems as well (SuSE, Ubuntu).

It's a quite minor issue, since I can't imagine many cases where there will be a distinction (although I can imagine some), but just to let you know.

Thanks again for all the great work!

**Comment by oxedions** [comment]
Dear @erkrali 

Apologies for the delay of my answer.
Ok, indeed, I totally bugged during the devs :blush: 

I reopen this issue and I will fix it asap.

**Comment by oxedions** [comment]
Dear @erkrali 

Here is the fix at last : https://github.com/bluebanquise/bluebanquise/pull/937/commits/2ee710acc05c330633127378220ad0c0fddecfeb

Really sorry for the delay, I had to manage 2 weeks of kids school holidays, then 1 week of kids sick :blush:

**Comment by oxedions** [comment]
Dear @erkrali 

I am closing this issue, all should be ok now 😊
